src/fsharp/FSharp.Core/resumable.fsi

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace Microsoft.FSharp.Core.CompilerServices
+
+#if !BUILDING_WITH_LKG && !BUILD_FROM_SOURCE
+open Microsoft.FSharp.Collections
+open Microsoft.FSharp.Core
+open System
+open System.Runtime.CompilerServices
+
+/// Acts as a template for struct state machines introduced by __stateMachine, and also as a reflective implementation
+[<Struct; NoComparison; NoEquality>]
+type ResumableStateMachine<'Data> =
+
+    /// When statically compiled, holds the data for the state machine
+    [<DefaultValue(false)>]
+    val mutable Data: 'Data
+
+    /// When statically compiled, holds the continuation goto-label further execution of the state machine
+    [<DefaultValue(false)>]
+    val mutable ResumptionPoint: int
+
+    /// Represents the delegated runtime continuation for a resumable state machine created dynamically
+    [<DefaultValue(false)>]
+    val mutable ResumptionDynamicInfo: ResumptionDynamicInfo<'Data>
+
+    interface IResumableStateMachine<'Data>
+
+    interface IAsyncStateMachine
+
+and IResumableStateMachine<'Data> =
+    /// Get the resumption point of the state machine
+    abstract ResumptionPoint: int
+
+    /// Copy-out or copy-in the data of the state machine
+    abstract Data: 'Data with get, set
```

Comment:
This is a good question.  My returning-byrefs foo may not be up-to-scratch.

This is evidently not allowed:
```fsharp
type IC =
    abstract Data: inref<int>

[<Struct>]
type C = 
    [<DefaultValue(false)>]
    val mutable a: int
    interface IC with
       member x.Data = &x.a // a.fs(8,25): error FS3232: Struct members cannot return the address of fields of the struct by reference
```
I'm not sure why this limitation applies.   The comment in the code says:
```
        // C# applies a rule where the APIs to struct types can't return the addresses of fields in that struct.
        // There seems no particular reason for this given that other protections in the language, though allowing
        // it would mean "readonly" on a struct doesn't imply immutability-of-contents
```

An alternative is this strange thing below where we pass in the 'this' pointer of the struct as an argument also:
```fsharp
type IC =
    abstract GetData: byref<C> -> byref<int>

and [<Struct>] C = 
    [<DefaultValue(false)>]
    val mutable a: int
    interface IC with
       member x.GetData(p) = &p.a
```

This would be enough - but is there a better way to get an interface to return a writable byref to a field of a struct?

Reply:
Right yes due to the lack of lifetime tracking returning a `this` ref can easily lead to memory safety bugs so it was disallowed entirely even though there are sound and valid cases like ours here.

```fs
[<Struct>]
type C = 
    [<DefaultValue(false)>]
    val mutable a: int
    member x.Data  = &x.a // say this did work

// oops
let foo: byref<int> = 
    let c = C()
    &c.Data
```

Indeed as your initial reply showed something passing in the reference would be needed to assure the compiler the lifetime is correctly managed by the outer scope. Here the example is an f-bounded interface but technically any C allowing GetData to produce 'Data would be fine of course.

```fs
type IC<'C,'Data when 'C :> IC<'C,'Data>> =
    abstract GetData: byref<'C> -> byref<'Data>

and [<Struct>] C<'Data> = 
    [<DefaultValue(false)>]
    val mutable a: 'Data
    interface IC<C<'Data>, 'Data> with
       member _.GetData(c) = &c.a
```

Anyway as I understand this interface is only used for the dynamic cases so it's not too important.

Reply:
Hmmm.... The interface is used for non-dynamic cases too, if the MoveNext needs to access the state machine.  The couroutines.fs example has a case of that, I think taskSeq.fs too

Will think it over

Reply:
What is the advantage of having a `byref` property?